proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,369 @@
+  3. Encode the string into the offset at memory
+```
+
+## `record.lift`
+
+The full syntax of this instruction is:
+```
+record.lift #typeidx
+```
+
+The type of this instruction is:
+```
+[$a, $b, ..] -> [$type]
+
+Where:
+  * $type is a type of the form (record $fields) taken from #typeidx
+  * $a, $b, .. are the #interface-valtype's for each field in $fields
+```
+
+The output is a record interface value with fields given by the inputs.
+
+## `record.lower`
+
+The full syntax of this instruction is:
+```
+record.lower #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [$fields]
+
+Where:
+  * $type is a valid type of the form (record $fields) taken from #typeidx
+```
+
+The outputs are the fields of the input `record`.
+
+## `variant.lift #typeidx $tag: #u32`
+
+The type of this instruction is:
+```
+[$payload] -> [$type]
+
+Where:
+  * $type is a type of the form (variant $options) taken from #typeidx
+  * $tag is an index of an (option #interface-valtype) from `$options`
+  * $payload = #interface-valtype
+```
+
+The output is a variant interface value with `tag` given by `$tag` and `payload` given by input.
+
+## `variant.lower_tags`
+
+The full syntax of this instruction is:
+```
+variant.lower_tag #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [i32]
+
+Where:
+ * $type is a valid type of the form (variant ..) taken from #typeidx
+```
+
+The output is the value of the `tag` of the input `variant`.
+
+## `variant.lower`
+
+This is a structured instruction with the following syntax:
+```
+variant.lower #typeidx $result: #interface-valtype*
+  (case $tag #interface-instr*)+
+end
+```
+
+The type of this instruction is:
+```
+[$type] -> [$result]
+
+Where:
+  * $type is a valid type of the form (variant $options) taken from #typeidx
+  * Each $tag uniquely matches an index of an option of $options
+  * Every option of $options is referenced by a case
+  * Each #interface-instr* is valid with the type [] -> [$result]
+```
+
+The output is the value of executed the matching `case` for the `tag` of the input `variant`.
+
+## `array.lift_memory`
+
+This is a structured instruction with the following syntax:
+```
+array.lift_memory #typeidx $width: #u32
+  interface-instr*
+end
+```
+
+The type of this instruction is:
+```
+[$base: i32, $length: i32] -> [$type]
+
+Where:
+  * $type is a type of the form (array $inner) taken from #typeidx
+  * #interface-instr* is valid with type [i32] -> [$inner]
+```
+
+This instruction will execute the following steps:
+```
+  1. Let $offsets = take([$base, $base + $width, $base + $width * 2, ..], $length)
+  2. Map #interface-instr* upon $offsets to yield the elements of the array
+```
+
+## `array.lower_memory`
+
+This is a structured instruction with the following syntax:
+```
+array.lower_memory #typeidx #memidx $alloc: #funcidx $width: #u32
+  interface-instr*
+end
+```
+
+The type of this instruction is:
+```
+[$type] -> []
```

Comment:
I think, dual to lifting, this instruction needs to yield 2 `i32`s (the offset and length)?

Reply:
Yes, good catch. I've added that.